pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -614,77 +617,62 @@ func (cm *containerManagerImpl) setFsCapacity() error {
+	// Allocate is thought to be called during predicate phase,
+	// just try to fetch device runtime information from cached state here
+	devOpts := cm.devicePluginHandler.GetDeviceRunContainerOptions(pod, container)
+	if devOpts == nil {
+		return nil, fmt.Errorf("Failed to get device run container options")
+	}
+	opts := &kubecontainer.RunContainerOptions{
+		Devices: devOpts.Devices,
+		Mounts:  devOpts.Mounts,
+		Envs:    devOpts.Envs,
+	}
+
+	return opts, nil
+}
+
+func (cm *containerManagerImpl) UpdatePluginResources(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
+	pod := attrs.Pod
+	pods := attrs.OtherPods
+	// TODO: determine whether to take initContainer into consideration
+	for _, container := range pod.Spec.Containers {
+		if devOpts := cm.devicePluginHandler.GetDeviceRunContainerOptions(pod, &container); devOpts != nil {
+			continue
+		}
+		// No runtime information found in cache, make allocation
+		if err := cm.devicePluginHandler.Allocate(pod, &container); err != nil {
+			return err
 		}
 
-		// Updates RunContainerOptions.Devices.
-		for _, dev := range resp.Devices {
-			if d, ok := devsMap[dev.ContainerPath]; ok {
-				glog.V(3).Infof("skip existing device %s %s", dev.ContainerPath, dev.HostPath)
-				if d != dev.HostPath {
-					glog.Errorf("Container device %s has conflicting mapping host devices: %s and %s",
-						dev.ContainerPath, d, dev.HostPath)
-				}
-				continue
-			}
-			devsMap[dev.ContainerPath] = dev.HostPath
-			opts.Devices = append(opts.Devices, kubecontainer.DeviceInfo{
-				PathOnHost:      dev.HostPath,
-				PathInContainer: dev.ContainerPath,
-				Permissions:     dev.Permissions,
-			})
-		}
-		// Updates RunContainerOptions.Mounts.
-		for _, mount := range resp.Mounts {
-			if m, ok := mountsMap[mount.ContainerPath]; ok {
-				glog.V(3).Infof("skip existing mount %s %s", mount.ContainerPath, mount.HostPath)
-				if m != mount.HostPath {
-					glog.Errorf("Container mount %s has conflicting mapping host mounts: %s and %s",
-						mount.ContainerPath, m, mount.HostPath)
-				}
-				continue
-			}
-			mountsMap[mount.ContainerPath] = mount.HostPath
-			opts.Mounts = append(opts.Mounts, kubecontainer.Mount{
-				Name:          mount.ContainerPath,
-				ContainerPath: mount.ContainerPath,
-				HostPath:      mount.HostPath,
-				ReadOnly:      mount.ReadOnly,
-				// TODO: This may need to be part of Device plugin API.
-				SELinuxRelabel: false,
-			})
+	}
+	// Update nodeInfo, to ensure NodeInfo.allocatableResource
+	// at least equal to the pod's requested capacity
+	newAllocatableResource := make(map[v1.ResourceName]int64)
+	newPodRequest := predicates.GetResourceRequest(pod)
+	for k, v := range newPodRequest.ScalarResources {
+		resource := string(k)
+		needed := int(v)
+		if cm.devicePluginHandler.ShouldHandle(resource, needed) {
+			newAllocatableResource[k] = v
 		}
 	}
-	return opts, nil
+	for _, existingPod := range pods {
```

Comment:
For simplicity and performance, can we skip counting other pods' resource requirements in nodeInfo allocatable?